src/timer.rs

```diff
@@ -0,0 +1,126 @@
+use core::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+    time::Duration,
+};
+use tokio::time::{sleep_until, Instant, Sleep};
+
+// # Compliance: 5.2
+// To prevent split votes in the first place, election timeouts are
+// chosen randomly from a fixed interval (e.g., 150–300ms).
+const MIN_REARM_DURATION: Duration = Duration::from_millis(150);
+const MAX_REARM_DURATION: Duration = Duration::from_millis(300);
+// Constant value to make tests deterministic
+const TEST_REARM_DURATION: Duration = Duration::from_millis(200);
+
+/// Used for awaiting timeouts.
+///
+/// To ensure Raft continues to make progress, the timer
+/// is automatically rearmed after it expires.
+#[derive(Debug)]
+pub struct Timer {
+    // The sleep future
+    sleep: Pin<Box<Sleep>>,
+}
+
+impl Timer {
+     fn new() -> Self {
+        let duration = Self::rearm_duration();
+        let expire = Instant::now() + duration;
+        let sleep = Box::pin(sleep_until(expire));
+        Timer {  sleep }
+    }
+
+    /// Check if the sleep future has completed.
+     fn poll_ready(&mut self, ctx: &mut Context) -> Poll<()> {
+        let poll = self.sleep.as_mut().poll(ctx);
+
+        if poll.is_ready() {
+            // rearm the timer so we continue to make progress
+            self.rearm();
+        }
+
+        poll
+    }
+
+    fn rearm(&mut self) {
+        let duration = Self::rearm_duration();
+        let expire = Instant::now() + duration;
```

Comment:
I think we may be loosing ticks here if we don't get polled for long. That may or may not be fine (it likely is), but should probably be documented as such somewhere to make it explicit. Also refer to https://docs.rs/tokio/1.40.0/tokio/time/enum.MissedTickBehavior.html

Reply:
Isnt MissedTickBehavior only for [Interval](https://docs.rs/tokio/1.40.0/tokio/time/fn.interval.html), here we are using [Sleep](https://docs.rs/tokio/1.40.0/tokio/time/fn.sleep.html).

The idea is to manually poll the Timer (which `impl Future` and calls down to `Sleep`) from our event loop. So we might miss ticks but its purely based on how we write our event loop.

Reply:
Yeah, but you are effectively building an interval here, so the same constraints apply. If we are getting polled late then `Instant::now()` may be reporting a value way in the future (compared to the previous expiration) already, I think.

Reply:
> So we might miss ticks but its purely based on how we write our event loop.

I am not sure that's all there is. The code is running as a process and depends on how you get scheduled (which is out of your control), no matter how the event loop is designed. So if you don't get CPU time for a minute, right now you'd just drop the ticks. I think it may be fine for the use case at hand, but would probably be best to be explicit about the intended behavior, that's all I am saying.

Reply:
I added some documentation to elaborate on Timeout's purpose and how to use it